File: src/test/java/io/javalin/TestResponse.java
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 *
 */

package io.javalin;

import com.mashape.unirest.http.HttpMethod;
import com.mashape.unirest.http.HttpResponse;
import io.javalin.newutil.BaseTest;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.util.Random;
import org.apache.commons.io.IOUtils;
import org.junit.Test;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.hasItem;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.core.IsCollectionContaining.hasItems;

public class TestResponse extends BaseTest {

    private String MY_BODY = ""
        + "This is my body, and I live in it. It's 31 and 6 months old. "
        + "It's changed a lot since it was new. It's done stuff it wasn't built to do. "
        + "I often try to fill if up with wine. - Tim Minchin";

    @Test
    public void test_resultString() throws Exception {
        app.get("/hello", ctx ->
            ctx.status(418)
                .result(MY_BODY)
                .header("X-HEADER-1", "my-header-1")
                .header("X-HEADER-2", "my-header-2"));
        HttpResponse<String> response = http.call(HttpMethod.GET, "/hello");
        assertThat(response.getStatus(), is(418));
        assertThat(response.getBody(), is(MY_BODY));
        assertThat(response.getHeaders().getFirst("X-HEADER-1"), is("my-header-1"));
        assertThat(response.getHeaders().getFirst("X-HEADER-2"), is("my-header-2"));
    }

    @Test
    public void test_resultStream() throws Exception {
        byte[] buf = new byte[65537]; // big and not on a page boundary
        new Random().nextBytes(buf);
        app.get("/stream", ctx -> ctx.result(new ByteArrayInputStream(buf)));
        HttpResponse<String> response = http.call(HttpMethod.GET, "/stream");

        ByteArrayOutputStream bout = new ByteArrayOutputStream();
        assertThat(IOUtils.copy(response.getRawBody(), bout), is(buf.length));
        assertThat(buf, equalTo(bout.toByteArray()));
    }

    @Test
    public void test_redirectInBefore() throws Exception {
        app.before("/before", ctx -> ctx.redirect("/redirected"));
        app.get("/redirected", ctx -> ctx.result("Redirected"));
        assertThat(http.getBody("/before"), is("Redirected"));
    }

    @Test
    public void test_redirectInExceptionMapper() throws Exception {
        app.get("/get", ctx -> {
            throw new Exception();
        });
        app.exception(Exception.class, (exception, ctx) -> ctx.redirect("/redirected"));
        app.get("/redirected", ctx -> ctx.result("Redirected"));
        assertThat(http.getBody("/get"), is("Redirected"));
    }

    @Test
    public void test_redirect() throws Exception {
        app.get("/hello", ctx -> ctx.redirect("/hello-2"));
        app.get("/hello-2", ctx -> ctx.result("Redirected"));
        assertThat(http.getBody("/hello"), is("Redirected"));
    }

    @Test
    public void test_redirectWithStatus() throws Exception {
        app.get("/hello", ctx -> ctx.redirect("/hello-2", 301));
        app.get("/hello-2", ctx -> ctx.result("Redirected"));
        http.disableUnirestRedirects();
        assertThat(http.call(HttpMethod.GET, "/hello").getStatus(), is(301));
        http.enableUnirestRedirects();
        assertThat(http.call(HttpMethod.GET, "/hello").getBody(), is("Redirected"));
    }

    @Test
    public void test_redirectWithStatus_absolutePath() throws Exception {
        app.get("/hello-abs", ctx -> ctx.redirect(origin + "/hello-abs-2", 303));
        app.get("/hello-abs-2", ctx -> ctx.result("Redirected"));
        http.disableUnirestRedirects();
        assertThat(http.call(HttpMethod.GET, "/hello-abs").getStatus(), is(303));
        http.enableUnirestRedirects();
        assertThat(http.call(HttpMethod.GET, "/hello-abs").getBody(), is("Redirected"));
    }

    @Test
    public void test_createCookie() throws Exception {
        app.post("/create-cookies", ctx -> ctx.cookie("name1", "value1").cookie("name2", "value2"));
        assertThat(http.post("/create-cookies").asString().getHeaders().get("Set-Cookie"), hasItems("name1=value1", "name2=value2"));
    }

    @Test
    public void test_cookie() throws Exception {
        app.post("/create-cookie", ctx -> ctx.cookie("Test", "Tast"));
        assertThat(http.post("/create-cookie").asString().getHeaders().get("Set-Cookie"), hasItem("Test=Tast"));
    }

}

File: src/test/java/io/javalin/TestResponse.kt
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 *
 */

package io.javalin

import com.mashape.unirest.http.HttpMethod
import io.javalin.core.util.Header
import io.javalin.util.TestUtil
import org.apache.commons.io.FileUtils
import org.apache.commons.io.IOUtils
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import org.junit.Test
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.File
import java.util.*
import javax.servlet.http.Cookie

class TestResponse {

    private val MY_BODY = (""
            + "This is my body, and I live in it. It's 31 and 6 months old. "
            + "It's changed a lot since it was new. It's done stuff it wasn't built to do. "
            + "I often try to fill if up with wine. - Tim Minchin")

    @Test
    fun `setting a String result works`() = TestUtil.test { app, http ->
        app.get("/hello") { ctx ->
            ctx.status(418).result(MY_BODY).header("X-HEADER-1", "my-header-1").header("X-HEADER-2", "my-header-2")
        }
        val response = http.call(HttpMethod.GET, "/hello")
        assertThat(response.status, `is`(418))
        assertThat(response.body, `is`(MY_BODY))
        assertThat(response.headers.getFirst("X-HEADER-1"), `is`("my-header-1"))
        assertThat(response.headers.getFirst("X-HEADER-2"), `is`("my-header-2"))
    }

    @Test
    fun `setting an InputStream result works`() = TestUtil.test { app, http ->
        val buf = ByteArray(65537) // big and not on a page boundary
        Random().nextBytes(buf)
        app.get("/stream") { ctx -> ctx.result(ByteArrayInputStream(buf)) }
        val response = http.call(HttpMethod.GET, "/stream")
        val bout = ByteArrayOutputStream()
        assertThat(IOUtils.copy(response.rawBody, bout), `is`(buf.size))
        assertThat(buf, equalTo(bout.toByteArray()))
    }

    @Test
    fun `setting an InputStream result works and InputStream is closed`() = TestUtil.test { app, http ->
        val path = "src/test/my-file.txt"
        File(path).printWriter().use { out ->
            out.print("Hello, World!")
        }
        app.get("/file") { ctx ->
            ctx.result(FileUtils.openInputStream(File(path)))
        }
        assertThat(http.getBody("/file"), `is`("Hello, World!"))
        assertThat(File(path).delete(), `is`(true))
    }

    @Test
    fun `redirect in before-handler works`() = TestUtil.test { app, http ->
        app.before("/before") { ctx -> ctx.redirect("/redirected") }
        app.get("/redirected") { ctx -> ctx.result("Redirected") }
        assertThat(http.getBody("/before"), `is`("Redirected"))
    }

    @Test
    fun `redirect in exception-mapper works`() = TestUtil.test { app, http ->
        app.get("/get") { ctx -> throw Exception() }
        app.exception(Exception::class.java) { exception, ctx -> ctx.redirect("/redirected") }
        app.get("/redirected") { ctx -> ctx.result("Redirected") }
        assertThat(http.getBody("/get"), `is`("Redirected"))
    }

    @Test
    fun `redirect in normal handler works`() = TestUtil.test { app, http ->
        app.get("/hello") { ctx -> ctx.redirect("/hello-2") }
        app.get("/hello-2") { ctx -> ctx.result("Redirected") }
        assertThat(http.getBody("/hello"), `is`("Redirected"))
    }

    @Test
    fun `redirect with status works`() = TestUtil.test { app, http ->
        app.get("/hello") { ctx -> ctx.redirect("/hello-2", 301) }
        app.get("/hello-2") { ctx -> ctx.result("Redirected") }
        http.disableUnirestRedirects()
        assertThat(http.call(HttpMethod.GET, "/hello").status, `is`(301))
        http.enableUnirestRedirects()
        assertThat(http.call(HttpMethod.GET, "/hello").body, `is`("Redirected"))
    }

    @Test
    fun `redirect to absolute path works`() = TestUtil.test { app, http ->
        app.get("/hello-abs") { ctx -> ctx.redirect("${http.origin}/hello-abs-2", 303) }
        app.get("/hello-abs-2") { ctx -> ctx.result("Redirected") }
        http.disableUnirestRedirects()
        assertThat(http.call(HttpMethod.GET, "/hello-abs").status, `is`(303))
        http.enableUnirestRedirects()
        assertThat(http.call(HttpMethod.GET, "/hello-abs").body, `is`("Redirected"))
    }

    @Test
    fun `setting a cookie works`() = TestUtil.test { app, http ->
        app.get("/create-cookie") { ctx -> ctx.cookie("Test", "Tast") }
        app.get("/get-cookie") { ctx -> ctx.result(ctx.cookie("Test")!!) }
        assertThat(http.get("/create-cookie").headers.getFirst(Header.SET_COOKIE), `is`("Test=Tast;Path=/"))
        assertThat(http.getBody("/get-cookie"), `is`("Tast"))
    }

    @Test
    fun `setting a Cookie object works`() = TestUtil.test { app, http ->
        app.get("/create-cookie") { ctx -> ctx.cookie(Cookie("Hest", "Hast").apply { maxAge = 7 }) }
        assertThat(http.get("/create-cookie").headers.getFirst(Header.SET_COOKIE), containsString("Hest=Hast"))
        assertThat(http.get("/create-cookie").headers.getFirst(Header.SET_COOKIE), containsString("Max-Age=7"))
    }

}
